File: frontend/src/lib/api/portfolio.ts
import { api } from './client'

export interface MfImportBatchOut {
  id: string
  user_id: number
  statement_date: string | null
  investor_name: string | null
  pan: string | null
  source_filename: string
  total_holdings: number
  total_value: string
  created_at: string
}

export interface MfHoldingOut {
  id: string
  batch_id: string
  user_id: number
  amc_name: string
  scheme_name: string
  folio_number: string
  isin: string | null
  units: string
  nav: string
  current_value: string
  as_of_date: string | null
  created_at: string
}

export interface MfImportBatchDetail extends MfImportBatchOut {
  holdings: MfHoldingOut[]
}

export interface ManualAssetOut {
  id: string
  user_id: number
  asset_type: string
  name: string
  institution: string | null
  current_value: string
  purchase_value: string | null
  purchase_date: string | null
  maturity_date: string | null
  units: string | null
  unit_price: string | null
  notes: string | null
  created_at: string
  updated_at: string
}

export interface ManualAssetCreate {
  asset_type: string
  name: string
  institution?: string
  current_value: string
  purchase_value?: string
  purchase_date?: string
  maturity_date?: string
  units?: string
  unit_price?: string
  notes?: string
}

export interface ManualAssetUpdate {
  asset_type?: string
  name?: string
  institution?: string
  current_value?: string
  purchase_value?: string
  purchase_date?: string
  maturity_date?: string
  units?: string
  unit_price?: string
  notes?: string
}

export interface AssetBreakdownItem {
  asset_type: string
  label: string
  total_value: string
  count: number
}

export interface EquitySummary {
  current_value: string
  cost_basis: string
  unrealized_pnl: string
  position_count: number
}

export interface MfSummary {
  current_value: string
  holding_count: number
  last_imported: string | null
}

export interface ManualSummary {
  current_value: string
  count: number
  breakdown: AssetBreakdownItem[]
}

export interface NetWorthOut {
  equity: EquitySummary
  mutual_funds: MfSummary
  manual_assets: ManualSummary
  total_net_worth: string
  as_of: string
}

// ── CAS upload ─────────────────────────────────────────────────────────────────

export function uploadCas(file: File): Promise<MfImportBatchOut> {
  const form = new FormData()
  form.append('file', file)
  return api.postForm<MfImportBatchOut>('/portfolio/cas/upload', form)
}

export function listBatches(): Promise<MfImportBatchOut[]> {
  return api.get<MfImportBatchOut[]>('/portfolio/cas/batches')
}

export function getBatch(batchId: string): Promise<MfImportBatchDetail> {
  return api.get<MfImportBatchDetail>(`/portfolio/cas/batches/${batchId}`)
}

export function deleteBatch(batchId: string): Promise<void> {
  return api.delete<void>(`/portfolio/cas/batches/${batchId}`)
}

// ── Manual assets ──────────────────────────────────────────────────────────────

export function createAsset(payload: ManualAssetCreate): Promise<ManualAssetOut> {
  return api.post<ManualAssetOut>('/portfolio/assets', payload)
}

export function listAssets(assetType?: string): Promise<ManualAssetOut[]> {
  const qs = assetType ? `?asset_type=${assetType}` : ''
  return api.get<ManualAssetOut[]>(`/portfolio/assets${qs}`)
}

export function updateAsset(id: string, payload: ManualAssetUpdate): Promise<ManualAssetOut> {
  return api.put<ManualAssetOut>(`/portfolio/assets/${id}`, payload)
}

export function deleteAsset(id: string): Promise<void> {
  return api.delete<void>(`/portfolio/assets/${id}`)
}

// ── Net worth ──────────────────────────────────────────────────────────────────

export function getNetWorth(): Promise<NetWorthOut> {
  return api.get<NetWorthOut>('/portfolio/net-worth')
}

File: backend/app/schemas/portfolio.py
"""Pydantic schemas for Phase 11 — External Portfolio."""

from datetime import date, datetime
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, field_validator

MANUAL_ASSET_TYPES = frozenset({"gold", "fd", "ppf", "nps", "bonds", "real_estate", "other"})


# ── Manual Asset schemas ───────────────────────────────────────────────────────

class ManualAssetCreate(BaseModel):
    asset_type: str
    name: str
    institution: str | None = None
    current_value: Decimal
    purchase_value: Decimal | None = None
    purchase_date: date | None = None
    maturity_date: date | None = None
    units: Decimal | None = None
    unit_price: Decimal | None = None
    notes: str | None = None

    @field_validator("asset_type")
    @classmethod
    def validate_type(cls, v: str) -> str:
        if v not in MANUAL_ASSET_TYPES:
            raise ValueError(f"asset_type must be one of {sorted(MANUAL_ASSET_TYPES)}")
        return v

    @field_validator("current_value")
    @classmethod
    def validate_value(cls, v: Decimal) -> Decimal:
        if v < 0:
            raise ValueError("current_value must be non-negative")
        return v


class ManualAssetUpdate(BaseModel):
    asset_type: str | None = None
    name: str | None = None
    institution: str | None = None
    current_value: Decimal | None = None
    purchase_value: Decimal | None = None
    purchase_date: date | None = None
    maturity_date: date | None = None
    units: Decimal | None = None
    unit_price: Decimal | None = None
    notes: str | None = None

    @field_validator("asset_type")
    @classmethod
    def validate_type(cls, v: str | None) -> str | None:
        if v is not None and v not in MANUAL_ASSET_TYPES:
            raise ValueError(f"asset_type must be one of {sorted(MANUAL_ASSET_TYPES)}")
        return v

    @field_validator("current_value")
    @classmethod
    def validate_value(cls, v: Decimal | None) -> Decimal | None:
        if v is not None and v < 0:
            raise ValueError("current_value must be non-negative")
        return v


class ManualAssetOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    user_id: int
    asset_type: str
    name: str
    institution: str | None
    current_value: Decimal
    purchase_value: Decimal | None
    purchase_date: date | None
    maturity_date: date | None
    units: Decimal | None
    unit_price: Decimal | None
    notes: str | None
    created_at: datetime
    updated_at: datetime


# ── MF Holding / Batch schemas ─────────────────────────────────────────────────

class MfHoldingOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    batch_id: str
    user_id: int
    amc_name: str
    scheme_name: str
    folio_number: str
    isin: str | None
    units: Decimal
    nav: Decimal
    current_value: Decimal
    as_of_date: date | None
    created_at: datetime


class MfImportBatchOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    user_id: int
    statement_date: date | None
    investor_name: str | None
    pan: str | None
    source_filename: str
    total_holdings: int
    total_value: Decimal
    created_at: datetime


class MfImportBatchDetail(MfImportBatchOut):
    holdings: list[MfHoldingOut] = []


# ── Net-worth schemas ──────────────────────────────────────────────────────────

class AssetBreakdownItem(BaseModel):
    asset_type: str
    label: str
    total_value: Decimal
    count: int


class EquitySummary(BaseModel):
    current_value: Decimal
    cost_basis: Decimal
    unrealized_pnl: Decimal
    position_count: int


class MfSummary(BaseModel):
    current_value: Decimal
    holding_count: int
    last_imported: datetime | None


class ManualSummary(BaseModel):
    current_value: Decimal
    count: int
    breakdown: list[AssetBreakdownItem]


class NetWorthOut(BaseModel):
    equity: EquitySummary
    mutual_funds: MfSummary
    manual_assets: ManualSummary
    total_net_worth: Decimal
    as_of: datetime
